Fixed #29725 -- Removed unnecessary join in QuerySet.count() and exists() on a many to many relation.

MyungSeKyo opened the pull request:
https://code.djangoproject.com/ticket/29725

pope1ni commented:
@MyungSeKyo: I'd check out `django.test.utils.CaptureQueriesContext` and use that to ensure that there is no `INNER JOIN` in the SQL.

MyungSeKyo commented:
@pope1ni Thanks I will try to add a test by using that!

charettes commented:
From a quick investigation locally the optimization can only be turned on if
- `self.target_field.db_constraint`
- `connections[db].features.supports_foreign_key`

And `target_field=None` values must be excluded as well to account for nullable target relationships that happened to be excluded by the `INNER JOIN`.

MyungSeKyo commented:
@charettes 
Thanks for your advice.
I fixed the code and write the test according to your advice.

charettes commented on tests/many_to_many/tests.py:
```diff
@@ -554,3 +555,13 @@ def test_inherited_models_selects(self):
+
+        with CaptureQueriesContext(connection) as captured_query:
+            article.publications.count()
+            self.assertNotIn('INNER JOIN', captured_query[0]['sql'])
```
Might want to only test for `JOIN` presence as this wouldn't fail if `LEFT JOIN` was used.

charettes commented on tests/many_to_many/tests.py:
```diff
@@ -554,3 +555,13 @@ def test_inherited_models_selects(self):
+        article = self.a1
+
+        with CaptureQueriesContext(connection) as captured_query:
+            article.publications.count()
```
You could assert the `count()` value as well.

MyungSeKyo commented:
@charettes 
Sorry charettes. 
I tried to fix but i am not sure this is right because i don't understand certainly

charettes commented on tests/many_to_many/tests.py:
```diff
@@ -554,3 +555,17 @@ def test_inherited_models_selects(self):
+            self.assertNotIn('JOIN', captured_query[0]['sql'])
+
+            self.assertEqual(article.publications.exists(), True)
+            self.assertNotIn('JOIN', captured_query[1]['sql'])
```
Move the exists assertions to another test.

charettes commented on tests/many_to_many/tests.py:
```diff
@@ -554,3 +555,17 @@ def test_inherited_models_selects(self):
+            self.assertEqual(article.publications.count(), 1)
+            self.assertNotIn('JOIN', captured_query[0]['sql'])
+
+            self.assertEqual(article.publications.exists(), True)
```
`assertTrue`

charettes commented on django/db/models/fields/related_descriptors.py:
```diff
@@ -912,6 +912,30 @@ def get_prefetch_queryset(self, instances, queryset=None):
+            # If the through's target field's foreign integrity is enforced
+            # the EXISTS can be performed against the through table instead of
+            # INNER JOIN'ing the target table.
+            if self.target_field.db_constraint and not self.target_field.null:
```
Even if it's not tested above `self.target_field.null` should be considered as well in `count()` and a test added.

By the way, the optimization is still doable when `self.target_field.null`, you just have to make sure to include `filters['%s__isnull' % self.target_field_name] = False` in the manager filters.

charettes commented on django/db/models/fields/related_descriptors.py:
```diff
@@ -912,6 +912,30 @@ def get_prefetch_queryset(self, instances, queryset=None):
+            # If the through's target field's foreign integrity is enforced
+            # the COUNT can be performed against the through table instead of
+            # INNER JOIN'ing the target table.
+            if self.target_field.db_constraint:
```
The optimization should also be disabled `if not connections[db].features.supports_foreign_keys`.

MyungSeKyo commented:
@charettes 
I tried to fix code according to your advice.
Could you review again?

charettes commented on tests/many_to_many/tests.py:
```diff
@@ -554,3 +556,28 @@ def test_inherited_models_selects(self):
+
+            self.assertEqual(article.publications.count(), 1)
+            self.assertNotIn('JOIN', captured_query[1]['sql'])
+            self.assertIn('"publication_id" IS NOT NULL', captured_query[1]['sql'])
```
The fact only a single result is returned is a strong enough assertion here. Some database backend could translate `__isnull` to some different SQL.

charettes commented on tests/many_to_many/tests.py:
```diff
@@ -554,3 +556,28 @@ def test_inherited_models_selects(self):
+
+            self.assertTrue(article.publications.exists())
+            self.assertNotIn('JOIN', captured_query[3]['sql'])
+            self.assertIn('"publication_id" IS NOT NULL', captured_query[3]['sql'])
```
Ditto, I'd remove.

MyungSeKyo commented:
@charettes 
I'm sorry to bother you charettes. I learned a lot from your kind reviews
I added code to check whether the optimization is turned off if `supports_foreign_keys` is false.
And removed unnecessary code.

charettes commented on tests/many_to_many/tests.py:
```diff
@@ -554,3 +560,27 @@ def test_inherited_models_selects(self):
+            self.assertTrue(nullable_target_article.publications.exists())
+
+        with mock.patch.object(connection.features, 'supports_foreign_keys', False):
+            self.assertIsNone(article.publications.constrained_target)
```
I would personally have broken this test in two (`test_count_join_optimization`, `test_exist_join_optimization`) and asserted against the results instead of testing the `constrained_target` implementation details but I guess this works fine.

MyungSeKyo replied:
Thanks to your reviews @charettes : )
But, I don't understand how to assert against the results.
Is it OK just to check that 'JOIN' is included in the `captured_query` when `supports_foreign_keys` is `False` in this context?
Or is there any better way?

charettes replied:
Oh I realize that asserting against the results is problematic given all the engines we're testing against support foreign keys. In this case yes, using the same `JOIN` check against `captured_query` should do!

MyungSeKyo replied:
Thanks @charettes 
I fixed the tests according to your feedbacks

MyungSeKyo replied:
Thanks @charettes 
I fixed the tests according to your advice.

MyungSeKyo commented:
Thanks @charettes
I fixed the tests according to your advice.

timgraham commented on tests/many_to_many/tests.py:
```diff
@@ -554,3 +564,38 @@ def test_inherited_models_selects(self):
+
+    def test_count_query_optimization(self):
+        """
+        #29725 - Calling count() on a many to many relation
```
```
"""
QuerySet.count() on a many-to-many relation doesn't include an
unnecessary JOIN.
"""
```
Ticket references should be reserved for obscure issues (not needed here, I think).

MyungSeKyo commented:
@timgraham 
Thank you for your review.
I separated the tests and created new tests class and changed to positional arguments
